it/multipart/src/test/java/com/linecorp/armeria/common/multipart/MultipartCollector.java

```diff
@@ -0,0 +1,287 @@
+                                              writing = false;
+                                              if (closing) {
+                                                  try {
+                                                      maybeCloseFileChannel();
+                                                      future.complete(path);
+                                                  } catch (IOException e) {
+                                                      bodyPartSubscription.cancel();
+                                                      future.completeExceptionally(e);
+                                                  }
+                                              } else {
+                                                  bodyPartSubscription.request(1);
+                                              }
+                                          });
+                                      }
+
+                                      @Override
+                                      public void failed(Throwable e, Void attachment) {
+                                          eventLoop.submit(() -> {
+                                              writing = false;
+                                              bodyPartSubscription.cancel();
+                                              future.completeExceptionally(e);
+                                          });
+                                      }
+                                  });
+            }
+
+            @Override
+            public void onError(Throwable t) {
+                assert fileChannel != null;
+                closing = true;
+                if (writing) {
+                    return;
+                }
+                try {
+                    maybeCloseFileChannel();
+                } catch (IOException e) {
+                    t.addSuppressed(e);
+                }
+                future.completeExceptionally(t);
+            }
+
+            @Override
+            public void onComplete() {
+                assert fileChannel != null;
+                closing = true;
+                if (writing) {
+                    return;
+                }
+                try {
+                    maybeCloseFileChannel();
+                    future.complete(path);
+                } catch (IOException e) {
+                    future.completeExceptionally(e);
+                }
+            }
+
+            private void maybeCloseFileChannel() throws IOException {
+                if (fileChannel.isOpen()) {
+                    fileChannel.close();
+                }
+            }
+        });
+        return future;
+    }
+
+    public CompletableFuture<CustomAggregatedMultipart> future() {
+        return future;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        multipartSubscription = s;
+        eventLoop = RequestContext.mapCurrent(RequestContext::eventLoop,
```

Comment:
Curious if we can guarantee that `onSubscribe` runs with `RequestContext`? I used `eventLoop` to guarantee all the callback of nested `BodyPart` subscribe can run on the same thread.

Reply:
As you did, it is only guaranteed when a context-aware executor is specified.